Remove error return on !exists in validateMountConfigReg

cdhunt opened the pull request:
We found while [developing](https://gitlab.com/gitlab-org/gitlab-runner/merge_requests/706) Docker Executor for Windows support for GitLab CI that the CI Runner code takes advantage of the docker daemon behavior to create directories on the host when mounting volumes. With Windows Containers, this throws an error, however, the logic to create the missing directory is platform agnostic.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Removed the hard check if the directory exists on the host in mount config validation.
**- How I did it**
Changed the condition logic.
**- How to verify it**
Mount a volume to a directory that does not exist on the host. It should be created and not throw an error.
**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Fix the inconsistent behavior between Linux and Windows containers when mounting to a directory that does not already exist.

thaJeztah commented:
I seem to recall auto-creation was only kept around on Linux for backward compatibility (it was deprecated for a long time, but was too much of a breaking change to remove), however, on Windows it never worked, and because of that (at least at the time) it was decided to not enable it.

Some background in https://github.com/moby/moby/pull/19953, https://github.com/moby/moby/pull/21666, https://github.com/moby/moby/pull/16349

/cc @cpuguy83

cdhunt commented:
Thanks, @thaJeztah. I think I concur with a lot of the sentiment in #21666 (for both sides). Specifically, we're trying to avoid interacting with the host directly and only use the Docker API. I'm pretty sure the GitLab CI Runner is already largely dependent on this feature for Linux hosts. I have validated the create logic does work on Server 2016 build `14393.1358.amd64fre.rs1_release.170602-2252`.

taylorb-microsoft commented:
It seems like this behavior is here to stay on the Linux side (for good/for bad/for otherwise) for me parity across platforms is a really valuable principal especially as we roll out LCOW.  The original PR ([Moby: 19953](https://github.com/moby/moby/pull/19953)) on listed the motivation for the change as "it is not reasonable.in fact,we can skip non-existant volume host path and remove "MkdirAll" from Setup.".  Which doesn't give me much to go on as to why auto-creation is the wrong policy.  

It seems this also tied into a change was made in runC ([runc: 493](https://github.com/opencontainers/runc/pull/492)).  I can see a very reasonable position that runC fail in this case vs auto creation but that Docker or even ContainerD could support the auto creation of the directory as a management policy in effect.

All that to say without an good reason to depart from parity with Linux I would be in favor of accepting this (or a similar) change.

cc: @jhowardmsft @darrenstahlmsft

cpuguy83 commented:
I agree parity would be good here so long as it works.

PatrickLang commented:
pinging myself

thaJeztah commented:
We discussed this in the maintainers meeting, and we're okay with lifting this restriction for Windows

ping @cdhunt can you update the PR (remove the commented-out code), so that we can move this forward?

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "master" git@github.com:cdhunt/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842353698296
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

thaJeztah commented on volume/volume_test.go:
```diff
@@ -180,16 +179,15 @@ func TestParseMountRaw(t *testing.T) {
-			`c:\:/foo:xyzzy`:                     "invalid volume specification: ",
-			`/`:                                  "destination can't be '/'",
-			`/..`:                                "destination can't be '/'",
-			`c:\notexist:/foo`:                   `source path does not exist`,
```
Do we have a "positive" test for this? (i.e. this test-case should now pass, correct?)

cdhunt replied:
Code wise, there's nothing new to test. The coverage is the same either way. I did change the functional test to validate no error is thrown. Do you still think we should include the positive tests?

thaJeztah replied:
Well, we previously tested that this should fail, but I couldn't find a test in the Windows `valid` list that now verifies that the test passes with a non-existing path (but I may have overlooked)

cdhunt replied:
I've added `valid` tests for Windows and LCOW.

jhowardmsft commented:
Sorry, trying to catch up with my backlog. I guess my only concern with this is that docker users can now arbitrarily create folders on the host through the docker API which they can't necessarily clean up. Is that not a concern?

cpuguy83 commented:
This is pretty consistent with linux.

thaJeztah commented:
> users can now arbitrarily create folders on the host through the docker API 

Correct; it's not ideal, but it's what a lot of people use. We are recommending the `--mount` option, which doesn't have this behavior.

> which they can't necessarily clean up

Can they be removed from a privileged/elevated session?


@cdhunt looks like this needs a rebase; also looks like there's a commit that isn't signed off

cdhunt commented:
@thaJeztah I've rebased and it looks like all commits are signed.

jhowardmsft commented:
> This is pretty consistent with linux.

@cpuguy83 Right, but that wasn't what I was asking. I was asking if it's the *right* thing to be doing.

BMOD89 commented:
Is there any progress regarding this topic?

thaJeztah commented:
(Looks like it needs a rebase now)

@jhowardmsft @johnstep is this change a blocker for you?

jhowardmsft commented:
Not a blocker, but I have voiced my concern with it above.

johnstep commented:
While I disagree with this loophole in the API design, I think consistency with Linux is important. Cleaning up directories created this way may require an administrator, similar to how root can be required on Linux.

AndiAnS commented:
I agree. Also because of portability/compatibility issues for tools which should run in both container types.

cdhunt commented:
The file/folder structure for `volume` has changed since this PR was created. It's going to take more than a simple Rebase to get it back to were he can be merged. I'll try to do that this week, but I haven't looked to see if any of the code has changed.

thaJeztah commented:
@cdhunt I suspect the change was https://github.com/moby/moby/pull/36688

BMOD89 commented:
@cdhunt any news on this topic?

cdhunt commented:
Haven't had time yet.

cdhunt commented:
I have merged the changes into the new files.

cpuguy83 commented:
Looks like the merge did not quite work.

cdhunt commented:
I think it's in good order now.

cpuguy83 commented:
@cdhunt Can you squash your commits?

cdhunt commented:
@cpuguy83 Could you provide some guidance for that? When I try to squash 2 commits I end up with thousands of changes. I'm struggling to tame my messy branch history.

cpuguy83 commented:
`git rebase -i HEAD~5` and squash all down to the first commit.

thaJeztah commented:
There's some information about squashing here as well; https://docs.docker.com/v17.06/opensource/workflow/work-issue/#pull-and-rebase-frequently

thaJeztah commented:
@cdhunt thanks for squashing! Looks like there's two failures; looks related to this change;

```
18:34:26 --- FAIL: TestParseMountRawSplit (0.00s)
...
18:34:26 	parser_test.go:389: Expected error, was nil, for spec c:\foo\bar:\\.\pipe\foo
```

```
18:34:26 --- FAIL: TestValidateMount (0.00s)
18:34:26 	validate_test.go:58: expected "bind mount source path does not exist: c:\\foo", got %!q(<nil>), case: 8
```

cdhunt committed:
Remove error return on !exists

On Linux, volume binding creates the directory on the host. This conditional check on Windows was preventing Windows Containers from creating the directory producing inconsistent behavior between platforms.

Signed-off-by:  Chris Hunt <cdhunt@gmail.com>

thaJeztah commented:
no worries! Just noted that it was failing (github likely won't send a notification if that's happening).


Looks like one was fixed, but one is still failing; one down; one to go! 🎉 

```
14:40:23 --- FAIL: TestValidateMount (0.00s)
14:40:23 	validate_test.go:58: expected %!q(<nil>), got "invalid mount config for type \"bind\": bind mount source path does not exist: /foo", case: 8
```

cdhunt commented:
@thaJeztah Can you point me to that result? Or maybe more specifically the test automation. I'm pretty sure this is coming from the LCOW side and I'm not sure exactly how to run those tests.

thaJeztah commented:
> I'm pretty sure this is coming from the LCOW side and I'm not sure exactly how to run those tests.

CI in this repository currently doesn't run on Windows RS3 and up (so doesn't run LCOW), so likely not related to LCOW (perhaps there's a test-case that shouldn't be run on Linux though)

You can click the links next to the failing CI runs;

- janky (Linux amd64); https://jenkins.dockerproject.org/job/Docker-PRs/50444
- Windows RS1 (Windows Server 2016) https://jenkins.dockerproject.org/job/Docker-PRs-WoW-RS1/21753/console
- PowerPC https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/10872/console
- s390x ("z") https://jenkins.dockerproject.org/job/Docker-PRs-s390x/10739/console


Windows RS1 actually showed a panic in `TestContainersAPICreateMountsValidation`

```
17:28:48 PANIC: docker_api_containers_test.go:1706: DockerSuite.TestContainersAPICreateMountsValidation
17:28:48 
17:28:48 case 0
17:28:48 case 1
17:28:48 case 2
17:28:48 case 3
17:28:48 ... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x458341)
17:28:48 
17:28:48 d:/CI/CI-3285a6745/go/src/runtime/asm_amd64.s:573
17:28:48   in call32
17:28:48 d:/CI/CI-3285a6745/go/src/runtime/panic.go:502
17:28:48   in gopanic
17:28:48 d:/CI/CI-3285a6745/go/src/runtime/panic.go:63
17:28:48   in panicmem
17:28:48 d:/CI/CI-3285a6745/go/src/runtime/signal_windows.go:167
17:28:48   in sigpanic
17:28:48 docker_api_containers_test.go:1895
17:28:48   in DockerSuite.TestContainersAPICreateMountsValidation
17:28:48 d:/CI/CI-3285a6745/go/src/runtime/asm_amd64.s:573
17:28:48   in call32
17:28:48 d:/CI/CI-3285a6745/go/src/reflect/value.go:447
17:28:48   in Value.call
17:28:48 d:/CI/CI-3285a6745/go/src/reflect/value.go:308
17:28:48   in Value.Call
17:28:48 c:/gopath/src/github.com/docker/docker/vendor/github.com/go-check/check/check.go:816
17:28:48   in suiteRunner.forkTest.func1
17:28:48 c:/gopath/src/github.com/docker/docker/vendor/github.com/go-check/check/check.go:672
17:28:48   in suiteRunner.forkCall.func1
17:28:48 d:/CI/CI-3285a6745/go/src/runtime/asm_amd64.s:2361
17:28:48   in goexit
```

BMOD89 commented:
Any news regarding this topic?

thaJeztah commented:
@cdhunt were you still working on the test failures?

cdhunt commented:
@thaJeztah No, I'm not set up to develop and test the affected variations. Thank you for your help over the last 13 months.

thaJeztah commented:
No worries, thanks @cdhunt !

I'll have a look around as well if there's possibly someone that has time to carry this PR

thaJeztah commented:
@simonferquel @silvin-lubecki - perhaps of interest to you?

fastloris commented:
This is interesting to me. If I was to get the PR in a mergeable state would it be accepted?

thaJeztah commented:
@fastloris yes, definitely. I think the PR was almost done, but there were some complications to get the tests working correctly; if you're interested in carrying this, that would be appreciated 👍

thaJeztah commented:
Feel free to "ping" me if you need help (e.g. to restore/copy the branch in this PR; you can download the patch if needed using this URL; https://patch-diff.githubusercontent.com/raw/moby/moby/pull/35159.patch
